### DISCUSSION TITLE
Streaming response content from file object

### DISCUSSION THREAD

**Discussion by bra-fsn** [mitmproxy/mitmproxy#5277] [General]
I would like to serve local files from `mitmproxy`, but all I could implement is the following:
```python
def request(self, flow: http.HTTPFlow) -> None:
        with open("/tmp/testfile", "r+b") as f:
            flow.response = http.Response.make(
                200,
                f.read(),
                {"Content-Type": "application/octet-stream"}
            )
```

While this seems to work fine, it has some problems:

1. it reads the whole file into memory, which limits how large files can be returned
2. it seems to be leaking memory. If `testfile` is 1 GiB, each request will increase the `mitmproxy` memory usage by around 1G.

Ideally, the content in `make` arguments could be an open file object, which would be read through (without buffering the whole file into memory) and then closed after the request has been served, but according to the code this is not possible.
Is it possible to solve these two errors with the current code (using `8.0.0`) or do you have any plans to fix these in upcoming releases?
Thanks,

**Comment by mhils** [comment]
> it seems to be leaking memory. If testfile is 1 GiB, each request will increase the mitmproxy memory usage by around 1G.

If you use mitmdump, the memory should be dropped again once the file has been served completely.

Other than that this is nothing we plan on solving in the near future I'm afraid. It does not work very well with how our proxying logic works.

**Reply by bra-fsn** [reply]
No, I've used `mitmproxy -s script.py`.
Sad to hear that. :(

**Reply by bra-fsn** [reply]
@mhils 
Would it be (easily) possible to return a chunked response somehow? So instead of creating a response object with a huge `content` blob, either having a callback/iterator, which yields data until it finishes, or to make an iterable/chunkable response from the callbacks (like `request`), which would be able to return content?

For the first (sorry, might not be fully working, just to get the idea):
```python
class ChunkProvider:
    def __init__(self, filename):
        self.filename = filename
        self.f = open(filename, "rb")

    def read_data(self):
        """Called until returns None, "" etc."""
        data = self.f.read(16384)
        if not data:
            self.f.close()
        yield data

def request(self, flow: http.HTTPFlow) -> None:
        chunks = ChunkProvider("/tmp/file")
        flow.response = http.Response.make(
            200,
            chunks.read_data,
            {"Content-Type": "application/octet-stream"}
        )
```

For the second (optionally providing the content-length in `make_immediate`):
```python
    def request(self, flow: http.HTTPFlow) -> None:
        with (open("/tmp/file", "r") as f,
              http.Response.make_immediate(200, None, {"Content-Type": "application/octet-stream"}) as r):
            while True:
                data = f.read(16384)
                if not data:
                    break
                r.push_data(data)
```

**Reply by mhils** [reply]
Not really. The problem is that we don't have backpressure when writing (only when reading) at the moment, which is hard to fix architecturally. The best workaround I can think of is to spawn another (streaming) HTTP server and then enable streaming in mitmproxy and redirect there.

**Reply by bra-fsn** [reply]
Thanks a lot for your prompt answers!
I'm not really familiar with asyncio (the implementation, I've used Twisted and gevent a lot), but after taking a quick glance on the code, would it be that hard?
I see that now `data` (which would be a generator for local files) is passed here:
https://github.com/mitmproxy/mitmproxy/blob/d1d854f0126540b78bb3e4e4a627cfd3a7965cb5/mitmproxy/proxy/server.py#L222
and the socket is drained in the next line.
So instead of doing this, couldn't `server_event` iterate on the generator passed in `command.data` and `await` on `writer.drain()` here:
https://github.com/mitmproxy/mitmproxy/blob/d1d854f0126540b78bb3e4e4a627cfd3a7965cb5/mitmproxy/proxy/server.py#L305
?
Any problems with converting `server_event` an `async def` and then `await` for it everywhere (or iterate on the generator in other ways and `await` on `drain` in each iterations?
I made a quick try and seems to be working (a 1G file was served from a generator, backed by a local file, without increasing the memory usage), although the console was blocked and so the downloads (the second was waiting for the first), so it's not perfect yet. ;)
BTW, streaming and proxying to a cache server seems to be a viable option.
